client/webserver/webserver.go

```diff
@@ -564,6 +560,43 @@ func (s *WebServer) Connect(ctx context.Context) (*sync.WaitGroup, error) {
 	return &wg, nil
 }
 
+// setAddr updates the listening address in case a :0 was provided and adds a
+// new directive to connect-src of content security policy if webserver address
+// is unspecified which is a workaround for a Safari browser bug.
+func (s *WebServer) setAddr(l net.Listener) {
+	var addToCSP bool
+	// Get the actual address that the listener is using.
+	addr := l.Addr().String()
+	host, port, err := net.SplitHostPort(addr)
+	if err == nil && host != "" {
+		addr = host
+	} else {
+		// If SplitHostPort failed, IPv6 addresses may still have brackets.
+		addr = strings.Trim(addr, "[]")
+	}
+
+	ip := net.ParseIP(addr)
```

Comment:
Also, since this method is not so trivial, we'd want to unit test it, and that would be easiest if the input were not a `net.Listener`.  Only `l.Addr()` is used, so it's a bit pointless to make it accept a `net.Listener`.  I'd make it a vanilla function, not a method, and consider using the typed [`net.TCPAddress`](https://pkg.go.dev/net#TCPAddr) somehow to avoid redundant processing.  Creating the listener did all this work already, including splitting host:port, and parsing the IP.

Also, I cannot tell at a glance looking at this method how it deals with IPv6 in general (not just [::] or [::1]).  It should not allow _any_ IPv6 to enter the csp string.